File: concepts/03_errors_and_resources/04_contracts_assertions_and_failure_signaling/lua/test_02_configuration_and_runtime_boundaries.lua
-- Common question: which invalid states fail during loading versus execution?
-- Inputs: malformed syntax, undeclared 5.5 globals, type errors, and missing modules.
-- Observations: load-time diagnostics, runtime pcall results, and require search reports.
-- polyglot-family: errors_and_resources
-- polyglot-concept: contracts_assertions_and_failure_signaling
-- polyglot-related: languages/lua/language/03_scope_and_variables/
-- polyglot-related+: test_023_global_const_and_loop_declarations.lua

local t = require("support.assertions")

local syntax, syntax_error = load("local =", "syntax", "t")
t.equal(syntax, nil)
t.equal(type(syntax_error), "string")

local undeclared, declaration_error = load(
    "global allowed; return missing",
    "global-contract",
    "t",
    {}
)
t.equal(undeclared, nil)
t.equal(type(declaration_error), "string")

t.raises(function() return 1 + {} end)
t.raises(function() require("polyglot_module_that_does_not_exist") end)

t.done()

File: concepts/03_errors_and_resources/04_contracts_assertions_and_failure_signaling/ruby/test_02_configuration_and_runtime_boundaries.rb
# 共同问题：无效程序在哪个阶段失败，语法、加载和运行期错误怎样区分。
# 输入：语法错误、缺失 feature 和错误操作数；观察：独立语法检查状态、LoadError 与 TypeError。
# polyglot-family: errors_and_resources
# polyglot-concept: contracts_assertions_and_failure_signaling
# polyglot-related: languages/ruby/tooling_and_runtime/13_command_parser_and_runtime/
# polyglot-related+: test_097_command_line_loading_warnings_and_option_parsing.rb

require "assertions"
require "helpers"

A = PolyglotAssertions
H = PolyglotRubyHelpers

stdout, stderr, status = H.ruby_command("-c", "-e", "def broken(")
A.falsey(status.success?)
A.equal("", stdout)
A.matches(/syntax error/i, stderr)

A.raises(LoadError) { require "polyglot_feature_that_does_not_exist" }
A.raises(TypeError) { 1 + "1" }
A.raises(NoMethodError) { Object.new.polyglot_missing_method }

A.done
